File: PA2/Lex.c
/** Mengdi Wei
  * ID:1538074
  * PA2
  * Lex.c can use insertion sort to make the words in  lexicographic order.
  */
#include<stdio.h>
#include<stdlib.h>
#include<string.h>
#include"List.h"
#define MAX_LEN 1000
	  
	int main ( int argc, char* argv[]) {
		 FILE *in, *out;
     
 	   if( argc != 3 ){
      		printf("Usage: %s <input file> <output file>\n", argv[0]);
      		exit(EXIT_FAILURE);
   	}


   	in = fopen(argv[1], "r");
   	if( in==NULL ){
      		printf("Unable to read from file %s\n", argv[1]);
      		exit(EXIT_FAILURE);
   	}

    
   	out = fopen(argv[2], "w");
   	
   	if( out==NULL ){
      		printf("Unable to write to file %s\n", argv[2]);
      		exit(EXIT_FAILURE);
   	}
   	
     char word [MAX_LEN ] ;
        int number = 0;
   	 while ( fgets(word, MAX_LEN , in) != NULL) {
   	 	  number++;
   	 }
   	 char* strs[MAX_LEN];
   	 rewind(in);
     
	   int n = 0;
	  
	    
	  while(  fgets(word,MAX_LEN , in) != NULL) {
	  	int j = strlen(word)+1;
		  strs[n] = malloc( sizeof(char)*j );
		  
      strcpy(strs[n], word );
      n++;
 	  }
 	  
 	  
 	  //These is insertionsort
 	  List list = newList();
	  append(list, 0);
	    for (int i = 1; i<n; i++) {
	    	char* arr = strs[i];
	    if (strcmp(arr, strs[back(list)])>0) {
	  		append (list, i);
	  		continue;
	  	}

	  	 	if (strcmp(arr, strs[front(list)])<0) {
	  		prepend (list, i);
	  		continue;
	  	}
	 
	  	moveFront(list);
	  	
	  	 while (index(list)!=-1) {
	  	 	 if (strcmp(arr, strs[get(list)])<0) {
	  	 	 	 insertBefore(list, i);
	  	 	 	 break;
	  	 	 	} 
	  	 	 	moveNext(list);
	  	 	 	}
	  	   
	  	 	 
	  	 }
	  
	  moveFront(list);
	   while (index(list)>=0) {
	   	int n = get(list);
	   	fprintf(out, "%s \n", strs[n]);
	   	 moveNext(list);
	   }
	     
   	   fclose(in);
       fclose(out);
	     freeList(&list);//free the list
	     for(int i = 0; i<number; i++){
		   free(strs[i]);
	     }//free the char*
	      
	    return(0);
	
	}

File: PA1/Lex.java
/** Mengdi Wei
  * Id: 1538074
  * This is Lex.java which deal with the input
  * output and the insertionsort method
  */

import java.util.*;
import java.io.*;

public class Lex{
	
	 String[] strs;
	 List list = new List();
	 
	public static void main(String args[]) throws Exception {
	   Lex exe=new Lex();
	   exe.Exe(args);
	 }
	  
	   void Exe (String args[]) throws Exception {
	  
	     String str;
	     int count = 0;
	     if (args.length!=2) {
	     	System.out.println("Usage:FileIO infile outfile ");
	     	System.exit(1);
	     }
	     Scanner scan = new Scanner(new FileInputStream(args[0]));
	     Scanner in = new Scanner(new FileInputStream(args[0]));
	     FileWriter out = new FileWriter (args[1]);
	     strs = new String[numbers(count,scan)];
	     int i =0;
	      while(in.hasNextLine()) {
	      	str=in.nextLine();
	      	strs[i]=str;
	        i++;
	      }
	      sort(list);
	     while (list.front!=null) {
	     	 out.write(strs[list.front()]);
	     	 out.write("\n");
	     	 list.front=list.front.next;
	     }
	     in.close();
	     scan.close();
	     out.close();
	   }
	   
	   //sort method can use insertion sort to make the lexicographic oder for the elements in the list 
	   void sort(List list) {
	  	 // System.out.println();
        int n = strs.length;
        list.append(0);
        String key;
        for (int i=1; i<n; i++) {
        	  
            key = strs[i];
            if (key.compareTo(strs[list.back()])>0) { 	
            	  list.append(i);
            	  continue;
            	  
            }  if (key.compareTo(strs[list.front()])<0) {
             	  list.prepend(i);
             	  continue;
             } 
             list.moveFront(); 
            while (list.index()!=-1 ) {
            	 if (key.compareTo(strs[list.get()])<0) {
            	 list.insertBefore(i);
            	 break;
            	} else {
            	list.moveNext();
             }
            }
            
           
        }
    }

    //numbers method can count how many lines in this file
	  int numbers (int count, Scanner scan) {
	  	while(scan.hasNextLine()) {
	  		 count++;
	  		// System.out.println(count);
	  		 String str = scan.nextLine();
	  	}
	  	return count;
	  }
	}

